src/lib/packager/clients/pkg_finish.rb

```diff
@@ -0,0 +1,171 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2016 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+#
+# ------------------------------------------------------------------------------
+
+require "installation/finish_client"
+require "packages/repository"
+
+module Yast
+  class PkgFinishClient < ::Installation::FinishClient
+    include Yast::I18n
+    include Yast::Logger
+
+    # Path to libzypp repositories
+    REPOS_DIR = "/etc/zypp/repos.d"
+    # Repository schemes to disable (disable_local_repos)
+    SCHEMES_TO_DISABLE = [:cd, :dvd]
+    # Path to failed_packages file
+    FAILED_PACKAGES_PATH = "/var/lib/YaST2/failed_packages"
+    # Command to create a tar.gz to back-up old repositories
+    TAR_CMD = "mkdir -p '%<source>s' && cd '%<source>s' "\
+      "&& /bin/tar -czf '%<archive>s' '%<target>s'"
+
+
+    # Constructor
+    def initialize
+      textdomain "packager"
+
+      Yast.import "Pkg"
+      Yast.import "Installation"
+      Yast.import "Mode"
+      Yast.import "Stage"
+      Yast.import "String"
+      Yast.import "FileUtils"
+      Yast.import "Packages"
+    end
+
+    # @see Implements ::Installation::FinishClient#modes
+    def modes
+      [:installation, :update, :autoinst]
+    end
+
+    # @see Implements ::Installation::FinishClient#title
+    def title
+      _("Saving the software manager configuration...")
+    end
+
+    # @see Implements ::Installation::FinishClient#write
+    def write
+      # Remove (backup) all sources not used during the update
+      # BNC #556469: Backup and remove all the old repositories before any Pkg::SourceSaveAll call
+      BackUpAllTargetSources() if Stage.initial && Mode.update
+
+      # See bnc #384827, #381360
+      if Mode.update
+        log.info("Adding default repositories")
+        WFM.call("inst_extrasources")
+      end
+
+      disable_uneeded_repos(SCHEMES_TO_DISABLE)
+
+      # save all repositories and finish target
+      Pkg.SourceSaveAll
+      Pkg.TargetFinish
+
+      # save repository metadata cache to the installed system
+      # (needs to be done _after_ saving repositories, see bnc#700881)
+      Pkg.SourceCacheCopyTo(Installation.destdir)
+
+      # copy list of failed packages to installed system
+      WFM.Execute(
```

Comment:
NTH: I'd prefer refactoring this and using the ``::File.exist?` and `::FileUtils.cp_a` from Ruby stdlib here.